### PR TITLE
Stack shuffling refactor

### PR DESCRIPTION
**Pull request by r0qs**

### CODE REVIEW

File: libyul/backends/evm/StackShuffle.h [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+		std::vector<int> multiplicity(m_numSlots, 0);
+		for (auto const& layoutSlot: m_stackLayout)
+			if (size_t const* slot = std::get_if<size_t>(&layoutSlot))
+			{
+				m_operationOutputs.insert(*slot);
+				--multiplicity[*slot];
+			}
+
+		for (auto&& slot: m_targetStack)
+			++multiplicity[slot];
```
Comment (r0qs) [inline libyul/backends/evm/StackShuffle.h]: Note that previously we were incrementing the multiplicity only if the following condition was satisfied:
```c++
if (outputs.count(slot) || generateSlotOnTheFly(slot))
```
That seems to be always the case (i.e. a slot in the target stack should necessarily be in the output or be able to be generated on the fly), so I removed the condition. I may be wrong though :)

**Comment by r0qs** [comment]
Current `develop` branch:
```
=======================================================
            chains.sol
-------------------------------------------------------
legacy pipeline took 0.70 seconds to execute.
via-ir pipeline took 84.76 seconds to execute.
=======================================================
=======================================================
            OptimizorClub.sol
-------------------------------------------------------
legacy pipeline took 3.24 seconds to execute.
via-ir pipeline took 17.09 seconds to execute.
=======================================================
=======================================================
            verifier.sol
-------------------------------------------------------
legacy pipeline took 0.52 seconds to execute.
via-ir pipeline took 2.81 seconds to execute.
=======================================================
```

This branch:
```
=======================================================
            chains.sol
-------------------------------------------------------
legacy pipeline took 0.68 seconds to execute.
via-ir pipeline took 62.68 seconds to execute.
=======================================================
=======================================================
            OptimizorClub.sol
-------------------------------------------------------
legacy pipeline took 3.12 seconds to execute.
via-ir pipeline took 17.14 seconds to execute.
=======================================================
=======================================================
            verifier.sol
-------------------------------------------------------
legacy pipeline took 0.51 seconds to execute.
via-ir pipeline took 2.68 seconds to execute.
=======================================================
```

Only `chains.sol` seems to have a significant performance improvement though.

File: libyul/backends/evm/StackShuffle.h (hunk 1 above)
Comment (r0qs) [reply libyul/backends/evm/StackShuffle.h]: It seems that my assumption was wrong after all, so I added back the condition: https://app.circleci.com/pipelines/github/ethereum/solidity/33813/workflows/ffa08b0d-dff6-4b99-8bee-841d32fb4fd6/jobs/1526831

**Comment by r0qs** [comment]
Benchmarks results:

- `develop` branch (commit [272892e9](https://github.com/ethereum/solidity/commit/272892e98e3f0f88be8aca063aad84008eadfad3)):

| File                 | Pipeline | Bytecode size | Time     | Exit code |
|----------------------|----------|--------------:|---------:|----------:|
| `verifier.sol`       | legacy   |    4874 bytes |   0.13 s |         0 |
| `verifier.sol`       | via-ir   |    4351 bytes |   0.59 s |         0 |
| `OptimizorClub.sol`  | legacy   |       0 bytes |   0.46 s |         1 |
| `OptimizorClub.sol`  | via-ir   |   22193 bytes |   3.38 s |         0 |
| `chains.sol`         | legacy   |    5845 bytes |   0.17 s |         0 |
| `chains.sol`         | via-ir   |   23043 bytes |  15.12 s |         0 |

- This branch:

| File                 | Pipeline | Bytecode size | Time     | Exit code |
|----------------------|----------|--------------:|---------:|----------:|
| `verifier.sol`       | legacy   |    4874 bytes |   0.14 s |         0 |
| `verifier.sol`       | via-ir   |    4351 bytes |   0.58 s |         0 |
| `OptimizorClub.sol`  | legacy   |       0 bytes |   0.47 s |         1 |
| `OptimizorClub.sol`  | via-ir   |   22193 bytes |   3.33 s |         0 |
| `chains.sol`         | legacy   |    5845 bytes |   0.17 s |         0 |
| `chains.sol`         | via-ir   |   23043 bytes |  12.15 s |         0 |

**Comment by cameel** [comment]
So only `chains.sol` is affected, just like in #14112? Odd. We need more benchmark contracts.

**Comment by cameel** [comment]
It would be interesting to see it benchmarked the way I did in https://github.com/ethereum/solidity/pull/14854#issuecomment-2070222067

This would show pure compilation time on some actual projects (I chose Uniswap v4 and Seaport since their latest versions compile without any workarounds) but with less variance than we see in CI. Or at least in a way that makes it easier to run multiple times and average the results.

**Comment by cameel** [comment]
On the topic of the PR itself - I tried to review it but the diff is almost useless due to all the moved code. It also has no description or changelog and commit descriptions are not very informative, so I'm not really sure what to expect here. The title sounds as if you were just making the code nicer but it does not seem to be all that you're doing here. If you could separate the refactors from actual functional changes, it would be much easier to review.

**Pull request closed without merging**